Fixed bugs found by the ShellCheck linter

tdulcet opened the pull request:
~**Edit**: This PR is now very outdated, but I would be happy to update it if there is ever interest in fixing these bugs.~

- Fixed numerous bugs found by [ShellCheck](https://www.shellcheck.net/).
  - See many of the bugs [here](https://travis-ci.org/tdulcet/mailinabox/builds/445985225#L1374) (lines 1374-2490).
  - Removed all the depreciated/legacy/antiquated Bash syntax (such as `$[..]` and `` `..` ``) and commands (such as `tempfile`).
- ~Outputs error messages to stderr.~ (Removed)
- Removed the use of many unnecessary commands.
  - Removed all the unnecessary `cat` and `echo` commands.
  - Replaced all the uses of the `pwd` command with Bash's `$PWD` variable.
- ~Fixed minor spelling/capitalization errors in script documentation.~ (Removed)

This pull request was originally part of #1435.

Closes #2342

zatricky commented on setup/start.sh:
```diff
@@ -35,7 +35,8 @@ if [ -f /etc/mailinabox.conf ]; then
 	# Load the old .conf file to get existing configuration options loaded
 	# into variables with a DEFAULT_ prefix.
-	cat /etc/mailinabox.conf | sed s/^/DEFAULT_/ > /tmp/mailinabox.prev.conf
+	cp /etc/mailinabox.conf /tmp/mailinabox.prev.conf
```
Doing this in two separate operations is weird.
```suggestion
	sed 's/^/DEFAULT_/' /etc/mailinabox.conf > /tmp/mailinabox.prev.conf
```

zatricky replied:
rm next line as well of course

tdulcet replied:
Did you test this? If so, great change and we can also do this below.

zatricky replied:
I'll get a chance to test properly this evening. It's a standard syntax I've used in hundreds of scripts though. :)

zatricky replied:
Tested. Same good end-result. :)

tdulcet replied:
Thank you! I will make the change here and at the two places below.

zatricky commented on setup/system.sh:
```diff
@@ -11,13 +11,12 @@ source setup/functions.sh # load our functions
 
-echo $PRIMARY_HOSTNAME > /etc/hostname
-hostname $PRIMARY_HOSTNAME
+hostnamectl set-hostname "$PRIMARY_HOSTNAME"
```
Not that we're worried about it for long - but I don't think this will work on Ubuntu 14.04?

tdulcet replied:
Yes, it does. It actually works on almost every currently supported Linux distribution.

zatricky replied:
The `hostnamectl` command is a part of systemd. I hadn't realised that, despite mostly using upstart, systemd-services are included in 14.04. I'm not certain if *all* 14.04 releases have it - but I don't think we'd be concerned, as long as it is in the latest one.

Good enough for me. :)

zatricky commented on setup/system.sh:
```diff
@@ -36,7 +35,7 @@ hostname $PRIMARY_HOSTNAME
 # for reference
 
-SWAP_MOUNTED=$(cat /proc/swaps | tail -n+2)
+SWAP_MOUNTED=$(< /proc/swaps tail -n+2)
```
This works - but for readability, I suggest:
```suggestion
SWAP_MOUNTED=$(tail -n+2 /proc/swaps)
```

tdulcet replied:
OK, good change!

zatricky commented on setup/system.sh:
```diff
@@ -328,15 +327,14 @@ restart_service resolvconf
 rm -f /etc/fail2ban/jail.local # we used to use this file but don't anymore
-cat conf/fail2ban/jails.conf \
-	| sed "s/PUBLIC_IP/$PUBLIC_IP/g" \
+< conf/fail2ban/jails.conf sed "s/PUBLIC_IP/$PUBLIC_IP/g" \
```
Given this is a multi-line set of piped commands, readability is probably better with the original

tdulcet replied:
Maybe, but it is also an error: https://github.com/koalaman/shellcheck/wiki/SC2002. Also, see [here](https://en.wikipedia.org/wiki/Cat_(Unix)#Useless_use_of_cat).

zatricky commented on setup/web.sh:
```diff
@@ -67,8 +67,7 @@ tools/editconf.py /etc/php/7.0/fpm/pool.d/www.conf -c ';' \
 chmod a+rx /var/lib/mailinabox
-cat conf/ios-profile.xml \
-	| sed "s/PRIMARY_HOSTNAME/$PRIMARY_HOSTNAME/" \
+< conf/ios-profile.xml sed "s/PRIMARY_HOSTNAME/$PRIMARY_HOSTNAME/" \
```
Same as 330 in system.sh - the cat is less efficient - but more readable

tdulcet replied:
I would suggest you run these two commands on your system from the mail-in-a-box repository:
`apt-get install shellcheck`
`bash -c 'shopt -s globstar; shellcheck -s bash **/*.sh'`
This will provide an explanation for all my changes in this pull request.

You can also see [here](https://travis-ci.org/tdulcet/mailinabox/builds/445985225#L1374) (lines 1374-2490).

zatricky replied:
100% true that the cat is *technically* useless and "wastes" a few CPU cycles and a few kilobytes of memory. But if we have to choose between that waste and having the script be easier to read/understand, I always go with it being easier to read. I defer to Josh's preferences.

zatricky commented on setup/web.sh:
```diff
@@ -81,18 +80,17 @@ chmod a+r /var/lib/mailinabox/mobileconfig.xml
 # and https://developer.mozilla.org/en-US/docs/Mozilla/Thunderbird/Autoconfiguration/FileFormat/HowTo.
-cat conf/mozilla-autoconfig.xml \
-	| sed "s/PRIMARY_HOSTNAME/$PRIMARY_HOSTNAME/" \
+< conf/mozilla-autoconfig.xml sed "s/PRIMARY_HOSTNAME/$PRIMARY_HOSTNAME/" \
```
In this case my preference is `sed in \n > out` or keeping the cat for consistency. Would defer to @JoshData.

tdulcet replied:
If you want to fix the bug, but keep it as two commands:
```
cp conf/mozilla-autoconfig.xml /var/lib/mailinabox/mozilla-autoconfig.xml
sed -i "s/PRIMARY_HOSTNAME/$PRIMARY_HOSTNAME/" /var/lib/mailinabox/mozilla-autoconfig.xml
```
Or, possibly your suggestion above: https://github.com/mail-in-a-box/mailinabox/pull/1457#discussion_r230141268.

zatricky commented on tools/owncloud-unlockadmin.sh:
```diff
@@ -4,20 +4,20 @@
+echo "If in doubt, press CTRL-C to cancel."
 echo 
-echo Press enter to continue.
+echo "Press enter to continue."
```
Perhaps take the opportunity to capitalise to `Enter`:
```suggestion
echo "Press Enter to continue."
```

tdulcet replied:
I was trying to keep this pull request to bugs found by ShellCheck and a few minor spelling errors, per your comment: https://github.com/mail-in-a-box/mailinabox/pull/1435#issuecomment-429839077.

zatricky commented:
Very good work overall. Lots of minor shell ambiguities and spelling issues rectified.

zatricky commented on setup/owncloud.sh:
```diff
@@ -9,7 +9,7 @@ source /etc/mailinabox.conf # load global vars
 echo "Installing Nextcloud (contacts/calendar)..."
 
-# Keep the php5 dependancies for the owncloud upgrades
+# Keep the php5 dependencies for the ownCloud upgrades
```
Small chance this is actually in reference to ownCloud's upgrade paths needed before migration to Nextcloud is possible?

tdulcet commented on setup/bootstrap.sh:
```diff
@@ -17,34 +17,34 @@ if [[ $EUID -ne 0 ]]; then
-		$HOME/mailinabox \
-		< /dev/null 2> /dev/null
+		"$HOME/mailinabox" \
+		< /dev/null
```
Yes, but it also suppresses any errors, causing `git` to fail silently.

tdulcet commented on setup/functions.sh:
```diff
@@ -3,25 +3,27 @@ function hide_output {
-		echo -----------------------------------------
-		cat $OUTPUT
-		echo -----------------------------------------
+		echo "FAILED: $*"
```
No, that would be an error: https://github.com/koalaman/shellcheck/wiki/SC2145.

zatricky replied:
It's an echo - pretty sure it results in the same thing.

tdulcet replied:
Yeah, echo is robust, although it is still incorrect to mix strings and arrays and ShellCheck will output that error...

tdulcet commented on setup/functions.sh:
```diff
@@ -3,25 +3,27 @@ function hide_output {
+		cat "$OUTPUT"
+		echo "-----------------------------------------"
+		# Remove temporary file.
+		rm -f "$OUTPUT"
```
No, because of the `exit` command below, the script exits and the temporary file is never removed.

tdulcet commented on setup/bootstrap.sh:
```diff
@@ -17,34 +17,34 @@ if [[ $EUID -ne 0 ]]; then
-		apt-get -q -q update
-		DEBIAN_FRONTEND=noninteractive apt-get -q -q install -y git < /dev/null
+		echo "Installing git . . ."
+		apt-get -qq update
```
No, because the `setup/functions.sh` file with the `hide_output` function, has not yet been downloaded. This file/script is installing `git` so it can clone the repository.

tdulcet commented:
@zatricky All your requested changes have been made in https://github.com/mail-in-a-box/mailinabox/pull/1457/commits/93fe6bf15313b74552fb72164237a10077a4a7fb.

JoshData commented:
Hey.

So, this is still too big for me to review. I'm also hesitant to spend time making changes that aren't (actual) bugs or new functionality. Every change has the potential to create new problem....

tdulcet commented:
There are plenty of "actual" bugs here. For example, I quoted all occurrences of the `STORAGE_ROOT` and `STORAGE_USER` variables, because the directories could contain a space, which would currently cause the script to fail. I removed all the depreciated/legacy/antiquated Bash syntax (such as `$[..]` and `` `..` ``) and Linux commands (such as `tempfile`), which could be removed at any time. I also removed many unnecessary `cat` and `echo` commands and made other changes, which simplify the scripts.

You can see the ShellCheck output [here](https://travis-ci.org/tdulcet/mailinabox/builds/445985225#L1374) (lines 1374-2490), which describes the most of the bugs in more detail.

bittorf commented:
We really should fix this:  
Maybe most of these problems are harmless, but having  
a chance to spot new problems is worth the effort.  

For now it looks like this:

```
#!/bin/sh
cd mailinabox-git
LIST="$( { find . -type f -iname '*.sh' && find . -type f -exec grep -l '^#!/bin/bash\|^#!/usr/bin/env sh' {} \; ;} | sort -u )"
for FILE in $LIST; do shellcheck -x "$FILE"; done | sed -n 's/.* \(SC[0-9][0-9][0-9][0-9]\) .*/\1/p' | sort | uniq -c | sort -rn

    192 SC2086
     17 SC1091
     11 SC2046
      7 SC2148
      7 SC2002
      6 SC2236
      6 SC2001
      3 SC2166
      3 SC2034
      2 SC2006
      1 SC2164
      1 SC2162
      1 SC2140
      1 SC2091
      1 SC2044
      1 SC2007
      1 SC2005
      1 SC2003
```

tdulcet commented:
> We really should fix this:
> Maybe most of these problems are harmless, but having
> a chance to spot new problems is worth the effort.

Thanks, I agree. I created this PR over 5 years ago now, but I would be happy to rebase and update it if there is a chance it could ever be merged.

In addition to those issues in the Bash scripts, there are also many bugs in the Python code, which can be found with the [Pylint](https://pylint.readthedocs.io/en/stable/) and [Ruff](https://docs.astral.sh/ruff/) linters among others. I could create a separate PR for this, if there is any interest in fixing these bugs... For example, here are the counts of the potential issues identified by Ruff (almost half of these are autofixable):
```
/mailinabox$ ruff --target-version py310 --select F,E4,E7,E9,W,I,UP,YTT,S,BLE,B,A,C4,T10,DJ,EM,EXE,ISC,ICN,G,PIE,PYI,Q003,Q004,RSE,RET,SLF,SLOT,SIM,TID,TCH,ARG,PGH,PL,TRY,FLY,PERF,FURB,LOG,RUF --preview --ignore W191,PLR09,PLR2004,RUF001,RUF002,RUF003 --line-length 320 . | head -n -2 | awk '{ print $2 }' | sort | uniq -c | sort -rn
error: Failed to parse tools/readable_bash.py:136:3: unindent does not match any outer indentation level
error: Failed to parse tools/mail.py:3:19: Unexpected token "$@"
     99 PLC0415
     85 E701
     47 UP031
     44 PLW1514
     37 I001
     33 W605
     32 PLR6201
     27 TRY003
     27 EM101
     26 PLC1901
     26 E401
     24 PLW2901
     22 UP032
     22 RET505
     20 FURB101
     19 UP015
     16 F401
     16 E722
     13 F841
     12 W293
     12 TRY200
     12 RUF005
     12 B904
     12 B007
     11 RSE102
     11 C401
     10 ARG001
      9 Q003
      8 SIM108
      8 E741
      8 BLE001
      7 TRY004
      7 TRY002
      6 SIM105
      6 SIM102
      6 S110
      6 RET504
      6 E711
      4 TRY300
      4 S603
      4 PERF401
      4 PERF203
      4 FURB113
      4 A002
      3 W291
      3 S607
      3 S324
      3 RET503
      3 PIE810
      2 UP030
      2 TRY301
      2 SIM110
      2 S310
      2 PLW0108
      2 ISC002
      2 G001
      2 F811
      2 E999
      2 E721
      2 E703
      1 W292
      1 UP041
      1 UP038
      1 UP034
      1 UP024
      1 SIM223
      1 SIM222
      1 SIM212
      1 SIM118
      1 SIM114
      1 S605
      1 S108
      1 RUF017
      1 RET507
      1 PLW0603
      1 PLW0120
      1 PLR6301
      1 PLR5501
      1 PLR1711
      1 PIE790
      1 ISC003
      1 G002
      1 F821
      1 EXE003
      1 EM103
      1 E713
      1 E712
      1 E402
      1 C414
      1 C405
      1 C404
      1 B012
      1 B006
      1 ARG005
      1 ARG002
      1 A001
```

JoshData commented:
My comment from 2018 still holds. It's hard for me to dedicate time to things that have no impact on functionality. The way to get the changes merged is to create smaller commits with one type of change per commit so that it's easy for me to understand what the changes are.

Also if you call something a bug but then don't demonstrate how Mail-in-a-Box operates incorrectly, it's a show-stopper for review, because I'm not going to do the work to investigate a "bug" that turns out to be a code style preference.

tdulcet commented:
Thanks Josh for your response. I can try to break this up into more commits, but note that many of the issues can now be fixed automatically by just running a single command like: `shellcheck -f diff -s bash **/*.sh | git apply`. Back in 2018 I had to make all of these fixes manually.

The term "bug" was used by ShellCheck to describe the type of issues that their linter detects, but I would be happy to use any alternative term that you prefer instead. For example, the term "error" is commonly used by other linters.

As I explained above back in 2018, I would consider almost all of the issues detected to be actual bugs, as either they address cases where the scripts would currently fail or they replace depreciated syntax/commands that could be removed and thus break at any time. Removing the depreciated syntax and commands should help significantly in preparation for the next Ubuntu LTS upgrade, as this most likely when it would otherwise break. A few of the changes removed unnecessary/useless commands to simply the script and improve readability/performance. If you consider this to be a "code style preference", I could remove these changes in the update/rebase. I also manually fixed a few typos I noticed in the script documentation, which I can remove as well, as they are obviously the least important of these fixes. I am not sure your definition of "code style preference", but I purposely did not include anything in this PR that I thought could be considered opinionated, although I would be happy to remove any of the individual changes if you feel differently. I would want to further hold up fixing the many critical issues this PR addresses.

I would highly encourage you to run ShellCheck yourself, as it explains each issue in more detail than I can in this PR. It also includes links to [their Wiki](https://www.shellcheck.net/wiki/Home), which explains each issue in even more detail, usually with examples and citations. For reference, here are the commands to run on Ubuntu and Debian:
```sh
sudo apt update
sudo apt install shellcheck
shopt -s globstar
shellcheck -s bash **/*.sh
```
(Note that this will miss a couple of scripts without the `.sh` extension. See @bittorf's `find` command above for an example of how to check those as well.)

bittorf commented:
This is just a starter - the question is, if this format  
(all fixes for one file) is OK and auditable and has  
a proper commit message.

tdulcet commented:
> The way to get the changes merged is to create smaller commits with one type of change per commit so that it's easy for me to understand what the changes are.

@JoshData - As requested, I redid this PR, doing only a single type of fix per commit. Two of the fixes are each broken up into two separate commits, so there are a total of 17 fixes and 19 commits. Most of the commits only modify a few lines or less.

Please let me know if you have any questions or if there is anything else you need me to do for this to be merged.

JoshData commented:
Ooops. Thanks. But this PR now has some conflicts with the main branch. Can you resolve those?

tdulcet commented:
No problem. Yes, I will rebase this PR from the main branch.

JoshData commented:
Thanks! I'm force-pushing to your branch to drop a few commits:

* [Fixed SC2140: Word is of the form "A"B"C".](https://github.com/mail-in-a-box/mailinabox/pull/1457/commits/1af77bf266f6bc93de0d49e8f6c237e521aa1a85): This commit was empty other than changing file permissions.
* [Fixed SC2002: Useless cat.](https://github.com/mail-in-a-box/mailinabox/pull/1457/commits/ce693c7981c61684fbc62d4a1074b0efa6d4893f): This makes it less readable.

I also squashed a few while I was here.

After I push I'll merge. Thanks for your patience!

bilogic commented:
May I also suggest standardizing on a code formatter?
